### PR TITLE
24374 Fix tf.einsum so it computes the trace correctly

### PR DESCRIPTION
**Pull request by lhendre**
This is in relation to and fixes issue #24374.  tf.einsum is not.  Einsum is not properly calculating Trace.  This fix functions by properly identifying a trace call and utilizing math_ops.trace.  Previously It would receive an error

`Subscript not supported: an axis appears more than once: i`
or something similar.  Before that error, people with older versions were receiving the wrongly calculated value.

### CODE REVIEW

File: tensorflow/python/ops/special_math_ops.py [hunk 1]
```diff
@@ -235,14 +234,14 @@ def einsum(equation, *inputs, **kwargs):
 
       output_axis_labels = ''.join(
           sorted(ax for ax in indices if counts[ax] == 1))
-
     for a in axis_labels:
       for input_labels in input_axis_labels:
+        if len(input_axis_labels) == 1 and input_labels.count(a) == 2 and input_labels==input_labels[::-1]:
```
Comment (drpngx) [inline tensorflow/python/ops/special_math_ops.py]: Can we update the doc?

Comment (lhendre) [reply tensorflow/python/ops/special_math_ops.py]: Updated the doc, let me know what you think

**Review by drpngx — DISMISSED** [review]
Thanks! I made one slight change to make the doc clearer, hopefully.

**Comment by drpngx** [comment]
Could you fix the lint errors?
```
tensorflow/python/ops/special_math_ops.py:242: [C0301(line-too-long), ] Line too long (107/80)
tensorflow/python/ops/special_math_ops.py:242: [C0326(bad-whitespace), ] Exactly one space required around comparison
tensorflow/python/ops/special_math_ops.py:243: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 10
```

**Comment by lhendre** [comment]
> Could you fix the lint errors?
> 
> ```
> tensorflow/python/ops/special_math_ops.py:242: [C0301(line-too-long), ] Line too long (107/80)
> tensorflow/python/ops/special_math_ops.py:242: [C0326(bad-whitespace), ] Exactly one space required around comparison
> tensorflow/python/ops/special_math_ops.py:243: [W0311(bad-indentation), ] Bad indentation. Found 12 spaces, expected 10
> ```

will do

**Comment by drpngx** [comment]
Did you test it? For some reason it seems to fail.
```
    _ = special_math_ops.einsum('ii->', x)
  File "/usr/lib/python3.4/unittest/case.py", line 181, in __exit__
    self._raiseFailure("{} not raised".format(exc_name))
  File "/usr/lib/python3.4/unittest/case.py", line 130, in _raiseFailure
    raise self.test_case.failureException(msg)
AssertionError: ValueError not raised
```

**Comment by lhendre** [comment]
Im in the process of building it right now, it takes awhile on my computer so I pushed to ensure i cant get any further feedback while i await the build, sorry for the confusion.  I will look into it

**Comment by lhendre** [comment]
I found the but related to that test, I have pushed the code up but I am still in testing.  My machine is taking some time to build so I pushed it in case I can get any early feedback

**Comment by lhendre** [comment]
The tests have appeared to have passed on my end, I havent had time to dig into all the logs but their appears to be no issues

**Comment by lhendre** [comment]
The tests passed on my end @drpngx

**Comment by lhendre** [comment]
So this is working on my end, I'll continue to tweak it to see if there are any better approaches while I await feedback.

**Comment by lhendre** [comment]
Thoughts?

**Comment by drpngx** [comment]
OK, let me run the tests again to see what's wrong.

**Comment by drpngx** [comment]
Did you test this test?

```
FAIL: test_simple (__main__.EinsumTest)
test_simple (__main__.EinsumTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/special_math_ops_test.runfiles/absl_py/absl/third_party/unittest3_backport/case.py", line 37, in testPartExecutor
    yield
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/special_math_ops_test.runfiles/absl_py/absl/third_party/unittest3_backport/case.py", line 162, in run
    testMethod()
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/special_math_ops_test.runfiles/org_tensorflow/tensorflow/python/ops/special_math_ops_test.py", line 292, in test_simple
    self.run_test(case)
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/special_math_ops_test.runfiles/org_tensorflow/tensorflow/python/ops/special_math_ops_test.py", line 359, in run_test
    self.assertLess(err, 1e-8)
AssertionError: 36.89950854256984 not less than 1e-08
```

**Comment by lhendre** [comment]
I believe I did?  Let me go back and check

**Comment by lhendre** [comment]
So not quite sure if I missed it or what but I was able to find and fix the error/bug.  The failing test was due to the added trace case('ijji').  The output is correct from the function, the issue is that the unit test does not correctly calculate the trace.  For now I removed the test and have left only one simple test case for trace('ii').  Let me know if that case is enough, otherwise, I can look into adding it back in and adding in code to ensure the unit test calculates the right value.  @drpngx

**Comment by lhendre** [comment]
2) Infra :: RabbitMQ :: Connection on connection should assign uri to connection:
     AssertionError: expected undefined to equal 'amqp://guest:guest@127.0.0.1:5672//?heartbeat=30'
      at Context.<anonymous> (test/unit/infra/messaging/RabbitMQConnect.spec.js:18:32)
      at process.topLevelDomainCallback (domain.js:121:23)

  3) Infra :: RabbitMQ :: Publishing Messages with a connection and no exchange defined should reject publish call with missing exchange:
     AssertionError: expected undefined to equal 'Publish failed - no exchange missing.ex on connection default is defined'
      at Context.it (test/unit/infra/messaging/RabbitMQPublishing.spec.js:24:34)

  4) Infra :: RabbitMQ :: Publishing Messages with a connection and exchange defined should not error on publish calls:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
  

  5) Infra :: RabbitMQ :: Publishing Messages with a connection and exchange defined "after all" hook:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
  

  6) Infra :: RabbitMQ :: Subscription "before all" hook:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
  

  7) utils :: emails when email is valid creates and sends the email:
     Error: Timeout of 5000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
  

  8) Socket.io :: Integration API :: POST /api/chats when sent data is ok should emit chat event (via socket) when a chat is added:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

**Comment by drpngx** [comment]
Yes, it would be nice to have that case work as well. Let me know when it's ready. Thanks!

**Comment by lhendre** [comment]
@drpngx I fixed the test and added it back in.  Thanks for the patience.  Its tested on my side and should be ready

**Comment by drpngx** [comment]
Thank you for the quick turnaround! Kicking off the tests...

**Comment by drpngx** [comment]
Could you fix the lint errors and reupload?

```
FAIL: Found 6 non-whitelited pylint errors:
tensorflow/python/ops/special_math_ops_test.py:355: [C0326(bad-whitespace), ] Exactly one space required around assignment
tensorflow/python/ops/special_math_ops_test.py:356: [C0326(bad-whitespace), ] Exactly one space required around comparison
tensorflow/python/ops/special_math_ops_test.py:357: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 6
tensorflow/python/ops/special_math_ops_test.py:357: [C0326(bad-whitespace), ] Exactly one space required after assignment
tensorflow/python/ops/special_math_ops_test.py:358: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 6
tensorflow/python/ops/special_math_ops_test.py:360: [W0311(bad-indentation), ] Bad indentation. Found 8 spaces, expected 6
```

**Comment by lhendre** [comment]
Yeah, sorry will fix those shortly

**Comment by lhendre** [comment]
Linting issues fixed @drpngx

**Comment by drpngx** [comment]
Testing...

**Comment by lhendre** [comment]
Tests run and things are linted on my side, not sure what is off with the internal builds

**Comment by drpngx** [comment]
Yeah the errors are unrelated. @ymodak it's ready.

**Comment by lhendre** [comment]
Yay!  Thanks for the review!

**Comment by drpngx** [comment]
Woohoo!

**Pull request merged**